## Data Clearning of Label Files (Not in project)

Get the binary version of OASIS dataset with unbalanced class distribution (More healthy subjects)   
Note that some row might have empty value for some features (We have limited data of AD)

### Processing the AD patients

In [8]:
import pandas as pd

# Load the AD subjects, igonoring their extent
# path = "/labels_AD.csv"
label_df = pd.read_csv(path)
label_df.head()

FileNotFoundError: [Errno 2] No such file or directory: '/labels_AD.csv'

In [2]:
label_df.describe()

SUBJECT#  Number of Time Points  Number of ROIs        MMSE  \
count   275.000000             275.000000           275.0  275.000000   
mean    668.403636             162.941818           132.0   25.694545   
std     367.624254               6.146944             0.0    3.798935   
min      26.000000              90.000000           132.0    9.000000   
25%     335.500000             164.000000           132.0   24.000000   
50%     660.000000             164.000000           132.0   26.000000   
75%     996.500000             164.000000           132.0   29.000000   
max    1324.000000             164.000000           132.0   30.000000   

       Age (at imaging)         CDR      CDR-SB        APOE         GDS  \
count        275.000000  275.000000  275.000000  275.000000  264.000000   
mean          75.167273    0.634545    2.852727   32.861818    3.090909   
std            7.444761    0.266922    2.185123    4.496856    5.872478   
min           50.000000    0.500000    0.500000   22.000000    0.000000   
25%           70.000000    0.500000    1.000000   33.000000    1.000000   
50%           76.000000    0.500000    2.500000   33.000000    2.000000   
75%           80.000000    0.500000    4.000000   34.000000    4.000000   
max           95.000000    2.000000   12.000000   44.000000   88.000000   

        education  ...  Disinhibition  Disinh_sev  Irritability   Irrit_sev  \
count  275.000000  ...     264.000000  264.000000    264.000000  264.000000   
mean    15.047273  ...       0.208333    0.329545      0.386364    0.609848   
std      3.047307  ...       0.406888    0.714674      0.487840    0.873815   
min      0.000000  ...       0.000000    0.000000      0.000000    0.000000   
25%     12.000000  ...       0.000000    0.000000      0.000000    0.000000   
50%     16.000000  ...       0.000000    0.000000      0.000000    0.000000   
75%     18.000000  ...       0.000000    0.000000      1.000000    1.000000   
max     20.000000  ...       1.000000    3.000000      1.000000    3.000000   

       Motor Disturbance  motdist_sev  nighttime behaviors   night_sev  \
count         264.000000   264.000000           264.000000  264.000000   
mean            0.140152     0.234848             0.261364    0.481061   
std             0.347804     0.627141             0.440212    0.990267   
min             0.000000     0.000000             0.000000    0.000000   
25%             0.000000     0.000000             0.000000    0.000000   
50%             0.000000     0.000000             0.000000    0.000000   
75%             0.000000     0.000000             1.000000    1.000000   
max             1.000000     3.000000             1.000000    9.000000   

         appetite  appeti_sev  
count  264.000000  264.000000  
mean     0.234848    0.371212  
std      0.424709    0.749198  
min      0.000000    0.000000  
25%      0.000000    0.000000  
50%      0.000000    0.000000  
75%      0.000000    0.000000  
max      1.000000    3.000000  

[8 rows x 35 columns]

In [3]:
print(label_df.shape)
print(label_df.columns.values)

(275, 40)
['SUBJECT#' 'Subject' 'MR ID' 'Number of Time Points' 'Number of ROIs'
 'Gender' 'MMSE' 'Age (at imaging)' 'CDR' 'CDR-SB' 'Primary Diagnosis'
 'APOE' 'GDS' 'education' 'hand' 'NPIQ Sum' 'Delusions' 'Delu_sev'
 'Hallucinations' 'Hall_sev' 'Agitation' 'Agit_sev' 'Depression'
 'Depres_sev' 'Anxiety' 'Anxie_sev' 'Elation' 'Elati_sev' 'Apathy'
 'Apath_sev' 'Disinhibition' 'Disinh_sev' 'Irritability' 'Irrit_sev'
 'Motor Disturbance' 'motdist_sev' 'nighttime behaviors' 'night_sev'
 'appetite' 'appeti_sev']


In [4]:
print(label_df['Primary Diagnosis'].unique())

['uncertain dementia' 'AD Dementia' 'Incipient Non-AD dem'
 'AD dem distrubed social- after' 'AD dem w/depresss- not contribut'
 'Unc: ques. Impairment' 'AD dem w/oth (list B) contribut' 'DLBD- primary'
 '0.5 in memory only' 'AD dem w/depresss- contribut'
 'Vascular Demt- primary' 'Frontotemporal demt. prim'
 'AD dem w/CVD not contrib' 'AD dem w/PDI after AD dem contribut' 'DAT'
 'uncertain- possible NON AD dem' 'AD dem w/oth (list B) not contrib'
 'Non AD dem- Other primary' 'AD dem Language dysf after'
 'AD dem Language dysf with' 'AD dem w/PDI after AD dem not contrib'
 'DLBD, primary' 'AD dem Language dysf prior' 'Dementia/PD- primary']


In [5]:
print(label_df['Primary Diagnosis'].describe())
print("-" * 50)
print(label_df['Primary Diagnosis'].value_counts())
print("-" * 50)
print(label_df[label_df.isnull().T.any()][['SUBJECT#', 'Subject','Primary Diagnosis']])
print("-" * 50)
check_df = label_df[['SUBJECT#', 'Subject','Primary Diagnosis']]
print(check_df[check_df.isnull().T.any()])

count             275
unique             24
top       AD Dementia
freq              133
Name: Primary Diagnosis, dtype: object
--------------------------------------------------
Primary Diagnosis
AD Dementia                              133
uncertain dementia                        84
DLBD- primary                              7
AD dem w/depresss- not contribut           6
0.5 in memory only                         5
AD dem w/depresss- contribut               5
uncertain- possible NON AD dem             4
Frontotemporal demt. prim                  4
AD dem Language dysf after                 3
AD dem w/oth (list B) not contrib          3
Vascular Demt- primary                     3
AD dem Language dysf with                  2
Non AD dem- Other primary                  2
AD dem w/CVD not contrib                   2
AD dem w/oth (list B) contribut            2
AD dem distrubed social- after             2
DAT                                        1
AD dem w/PDI after AD dem contribut    

For 'Vascular Demt- primary', 'Frontotemporal demt. prim', they might have different mechanism from AD.   
For '0.5 in memory only', it might be a kind of MCI. O means normal in a memory test.

In [6]:
AD_dropout_vals = ['uncertain dementia', 'Incipient Non-AD dem', 'Unc: ques. Impairment', 'DLBD- primary', \
                  'Vascular Demt- primary', 'Frontotemporal demt. prim', 'uncertain- possible NON AD dem', \
                  'DAT', 'uncertain- possible NON AD dem', 'Non AD dem- Other primary', 'DLBD, primary', \
                  'Dementia/PD- primary']

mask = label_df['Primary Diagnosis'].isin(AD_dropout_vals)
print(type(mask))
print(mask.shape) 
print(mask.head())
label_df = label_df.drop(label_df[mask].index)
print(label_df['Primary Diagnosis'].value_counts())
print("-" * 50)
print(label_df['Primary Diagnosis'].unique())
print(label_df.shape)

<class 'pandas.core.series.Series'>
(275,)
0     True
1     True
2    False
3     True
4     True
Name: Primary Diagnosis, dtype: bool
Primary Diagnosis
AD Dementia                              133
AD dem w/depresss- not contribut           6
0.5 in memory only                         5
AD dem w/depresss- contribut               5
AD dem w/oth (list B) not contrib          3
AD dem Language dysf after                 3
AD dem distrubed social- after             2
AD dem w/oth (list B) contribut            2
AD dem w/CVD not contrib                   2
AD dem Language dysf with                  2
AD dem w/PDI after AD dem contribut        1
AD dem w/PDI after AD dem not contrib      1
AD dem Language dysf prior                 1
Name: count, dtype: int64
--------------------------------------------------
['AD Dementia' 'AD dem distrubed social- after'
 'AD dem w/depresss- not contribut' 'AD dem w/oth (list B) contribut'
 '0.5 in memory only' 'AD dem w/depresss- contribut'
 'AD dem w/CVD

In [7]:
# print(label_df[mask].index.shape)
# print(label_df[mask].index)
# print(label_df.drop(label_df[mask].index))

In [8]:
# To make the label consistent!
print(label_df['Primary Diagnosis'].unique())
label_df['Primary Diagnosis'] = 'AD Dementia'
print(label_df['Primary Diagnosis'].unique())
`print(label_df.shape)`
print(label_df['Primary Diagnosis'].head())

['AD Dementia' 'AD dem distrubed social- after'
 'AD dem w/depresss- not contribut' 'AD dem w/oth (list B) contribut'
 '0.5 in memory only' 'AD dem w/depresss- contribut'
 'AD dem w/CVD not contrib' 'AD dem w/PDI after AD dem contribut'
 'AD dem w/oth (list B) not contrib' 'AD dem Language dysf after'
 'AD dem Language dysf with' 'AD dem w/PDI after AD dem not contrib'
 'AD dem Language dysf prior']
['AD Dementia']
(166, 40)
2     AD Dementia
5     AD Dementia
7     AD Dementia
10    AD Dementia
11    AD Dementia
Name: Primary Diagnosis, dtype: object


### Processing the healthy people

In [9]:
# Load the health subjects
hc_path = "OASIS/oasis_labels.csv"
hc_df = pd.read_csv(hc_path)
hc_df.head()

SUBJECT#   Subject              MR ID  Number of Time Points  \
0         1  OAS30001  OAS30001_MR_d0757                    164   
1         2  OAS30001  OAS30001_MR_d3132                    164   
2         3  OAS30002  OAS30002_MR_d2340                    164   
3         4  OAS30002  OAS30002_MR_d2345                    164   
4         5  OAS30003  OAS30003_MR_d1631                    164   

   Number of ROIs  Gender  MMSE  Age (at imaging)  CDR  CDR-SB  ...  \
0             132  female    30                67  0.0     0.0  ...   
1             132  female    30                73  0.0     0.0  ...   
2             132    male    30                73  0.0     0.0  ...   
3             132    male    30                73  0.0     0.0  ...   
4             132  female    30                63  0.0     0.0  ...   

  Disinhibition  Disinh_sev  Irritability  Irrit_sev Motor Disturbance  \
0           0.0         0.0           0.0        0.0               0.0   
1           0.0         0.0           0.0        0.0               0.0   
2           0.0         0.0           0.0        0.0               0.0   
3           0.0         0.0           0.0        0.0               0.0   
4           0.0         0.0           0.0        0.0               0.0   

   motdist_sev  nighttime behaviors  night_sev  appetite  appeti_sev  
0          0.0                  0.0        0.0       0.0         0.0  
1          0.0                  0.0        0.0       0.0         0.0  
2          0.0                  0.0        0.0       0.0         0.0  
3          0.0                  0.0        0.0       0.0         0.0  
4          0.0                  0.0        0.0       0.0         0.0  

[5 rows x 40 columns]

In [10]:
hc_df.describe()

SUBJECT#  Number of Time Points  Number of ROIs         MMSE  \
count  1326.000000            1326.000000     1326.000000  1326.000000   
mean    663.500000             163.518854      131.997738    28.377074   
std     382.927539               3.945958        0.061388     2.459247   
min       1.000000              90.000000      130.000000     9.000000   
25%     332.250000             164.000000      132.000000    28.000000   
50%     663.500000             164.000000      132.000000    29.000000   
75%     994.750000             164.000000      132.000000    30.000000   
max    1326.000000             164.000000      132.000000    30.000000   

       Age (at imaging)          CDR       CDR-SB         APOE          GDS  \
count       1326.000000  1326.000000  1326.000000  1326.000000  1277.000000   
mean          70.415535     0.131599     0.602941    32.307692     1.747063   
std            8.950516     0.284553     1.522584     4.491204     4.624226   
min           42.000000     0.000000     0.000000    22.000000     0.000000   
25%           65.000000     0.000000     0.000000    33.000000     0.000000   
50%           71.000000     0.000000     0.000000    33.000000     1.000000   
75%           76.000000     0.000000     0.000000    34.000000     2.000000   
max           95.000000     2.000000    12.000000    44.000000    88.000000   

         education  ...  Disinhibition   Disinh_sev  Irritability  \
count  1326.000000  ...    1277.000000  1277.000000   1277.000000   
mean     15.736802  ...       0.059514     0.090838      0.172279   
std       2.798806  ...       0.236678     0.389380      0.377771   
min       0.000000  ...       0.000000     0.000000      0.000000   
25%      14.000000  ...       0.000000     0.000000      0.000000   
50%      16.000000  ...       0.000000     0.000000      0.000000   
75%      18.000000  ...       0.000000     0.000000      0.000000   
max      24.000000  ...       1.000000     3.000000      1.000000   

         Irrit_sev  Motor Disturbance  motdist_sev  nighttime behaviors  \
count  1277.000000        1277.000000  1277.000000          1277.000000   
mean      0.249021           0.040720     0.067345             0.093187   
std       0.607433           0.197719     0.352119             0.290808   
min       0.000000           0.000000     0.000000             0.000000   
25%       0.000000           0.000000     0.000000             0.000000   
50%       0.000000           0.000000     0.000000             0.000000   
75%       0.000000           0.000000     0.000000             0.000000   
max       3.000000           1.000000     3.000000             1.000000   

         night_sev     appetite   appeti_sev  
count  1277.000000  1277.000000  1277.000000  
mean      0.155834     0.099452     0.145654  
std       0.564068     0.299385     0.482376  
min       0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000  
max       9.000000     1.000000     3.000000  

[8 rows x 35 columns]

In [11]:
print(hc_df.shape)
print(hc_df.columns.values)

(1326, 40)
['SUBJECT#' 'Subject' 'MR ID' 'Number of Time Points' 'Number of ROIs'
 'Gender' 'MMSE' 'Age (at imaging)' 'CDR' 'CDR-SB' 'Primary Diagnosis'
 'APOE' 'GDS' 'education' 'hand' 'NPIQ Sum' 'Delusions' 'Delu_sev'
 'Hallucinations' 'Hall_sev' 'Agitation' 'Agit_sev' 'Depression'
 'Depres_sev' 'Anxiety' 'Anxie_sev' 'Elation' 'Elati_sev' 'Apathy'
 'Apath_sev' 'Disinhibition' 'Disinh_sev' 'Irritability' 'Irrit_sev'
 'Motor Disturbance' 'motdist_sev' 'nighttime behaviors' 'night_sev'
 'appetite' 'appeti_sev']


In [12]:
print(hc_df['Primary Diagnosis'].unique())

['Cognitively normal' 'uncertain dementia' 'AD Dementia'
 'Incipient Non-AD dem' 'AD dem distrubed social- after'
 'AD dem w/depresss- not contribut' 'Unc: ques. Impairment'
 'AD dem w/oth (list B) contribut' 'DLBD- primary' '0.5 in memory only'
 'AD dem w/depresss- contribut' 'Vascular Demt- primary'
 'Frontotemporal demt. prim' 'AD dem w/CVD not contrib'
 'AD dem w/PDI after AD dem contribut' 'DAT'
 'uncertain- possible NON AD dem' 'AD dem w/oth (list B) not contrib'
 'No dementia' 'Non AD dem- Other primary' 'AD dem Language dysf after'
 'AD dem Language dysf with' 'AD dem w/PDI after AD dem not contrib'
 'DLBD, primary' 'AD dem Language dysf prior' '0' 'Dementia/PD- primary']


In [13]:
print(hc_df['Primary Diagnosis'].describe())
print("-" * 50)
print(hc_df['Primary Diagnosis'].value_counts())
print("-" * 50)
print(hc_df[hc_df.isnull().T.any()][['SUBJECT#', 'Subject','Primary Diagnosis']])
print("-" * 50)
hc_check_df = hc_df[['SUBJECT#', 'Subject','Primary Diagnosis']]
print(hc_check_df[hc_check_df.isnull().T.any()])

count                   1326
unique                    27
top       Cognitively normal
freq                    1048
Name: Primary Diagnosis, dtype: object
--------------------------------------------------
Primary Diagnosis
Cognitively normal                       1048
AD Dementia                               133
uncertain dementia                         84
DLBD- primary                               7
AD dem w/depresss- not contribut            6
0.5 in memory only                          5
AD dem w/depresss- contribut                5
uncertain- possible NON AD dem              4
Frontotemporal demt. prim                   4
AD dem Language dysf after                  3
Vascular Demt- primary                      3
AD dem w/oth (list B) not contrib           3
AD dem Language dysf with                   2
Non AD dem- Other primary                   2
No dementia                                 2
AD dem w/CVD not contrib                    2
AD dem w/oth (list B) contribut         

In [14]:
hc_vals = ['Cognitively normal', 'No dementia', '0'] # 'Non AD dem- Other primary', 'Incipient Non-AD dem'
hc_mask = hc_df['Primary Diagnosis'].isin(hc_vals)
hc_df = hc_df[hc_mask]
print(hc_df['Primary Diagnosis'].value_counts())
print("-" * 50)
print(hc_df['Primary Diagnosis'].unique())
print(hc_df.shape)

Primary Diagnosis
Cognitively normal    1048
No dementia              2
0                        1
Name: count, dtype: int64
--------------------------------------------------
['Cognitively normal' 'No dementia' '0']
(1051, 40)


In [15]:
# To make the label consistent!
print(hc_df['Primary Diagnosis'].unique())
hc_df['Primary Diagnosis'] = 'Cognitively normal'
print(hc_df['Primary Diagnosis'].unique())
print(hc_df.shape)
print(hc_df['Primary Diagnosis'].head())

['Cognitively normal' 'No dementia' '0']
['Cognitively normal']
(1051, 40)
0    Cognitively normal
1    Cognitively normal
2    Cognitively normal
3    Cognitively normal
4    Cognitively normal
Name: Primary Diagnosis, dtype: object


### Combine the healthy people and patients

In [16]:
print(label_df.shape)
print(label_df.head())
print("-" * 50)
print(hc_df.shape)
hc_df.head()

(166, 40)
    SUBJECT#   Subject              MR ID  Number of Time Points  \
2         38  OAS30027  OAS30027_MR_d2394                    164   
5         58  OAS30043  OAS30043_MR_d0145                    164   
7         71  OAS30051  OAS30051_MR_d1286                    164   
10        98  OAS30076  OAS30076_MR_d1735                    164   
11       108  OAS30085  OAS30085_MR_d1566                    164   

    Number of ROIs  Gender  MMSE  Age (at imaging)  CDR  CDR-SB  ...  \
2              132    male    26                75  0.5     1.5  ...   
5              132    male    24                86  1.0     5.0  ...   
7              132  female    30                69  0.5     3.0  ...   
10             132  female    23                76  0.5     3.5  ...   
11             132    male    30                70  0.5     3.0  ...   

   Disinhibition  Disinh_sev  Irritability  Irrit_sev Motor Disturbance  \
2            0.0         0.0           0.0        0.0               0.0  

SUBJECT#   Subject              MR ID  Number of Time Points  \
0         1  OAS30001  OAS30001_MR_d0757                    164   
1         2  OAS30001  OAS30001_MR_d3132                    164   
2         3  OAS30002  OAS30002_MR_d2340                    164   
3         4  OAS30002  OAS30002_MR_d2345                    164   
4         5  OAS30003  OAS30003_MR_d1631                    164   

   Number of ROIs  Gender  MMSE  Age (at imaging)  CDR  CDR-SB  ...  \
0             132  female    30                67  0.0     0.0  ...   
1             132  female    30                73  0.0     0.0  ...   
2             132    male    30                73  0.0     0.0  ...   
3             132    male    30                73  0.0     0.0  ...   
4             132  female    30                63  0.0     0.0  ...   

  Disinhibition  Disinh_sev  Irritability  Irrit_sev Motor Disturbance  \
0           0.0         0.0           0.0        0.0               0.0   
1           0.0         0.0           0.0        0.0               0.0   
2           0.0         0.0           0.0        0.0               0.0   
3           0.0         0.0           0.0        0.0               0.0   
4           0.0         0.0           0.0        0.0               0.0   

   motdist_sev  nighttime behaviors  night_sev  appetite  appeti_sev  
0          0.0                  0.0        0.0       0.0         0.0  
1          0.0                  0.0        0.0       0.0         0.0  
2          0.0                  0.0        0.0       0.0         0.0  
3          0.0                  0.0        0.0       0.0         0.0  
4          0.0                  0.0        0.0       0.0         0.0  

[5 rows x 40 columns]

In [17]:
label_df = pd.concat([label_df, hc_df],axis=0)
print(label_df.shape)

(1217, 40)


In [18]:
label_df.head()

SUBJECT#   Subject              MR ID  Number of Time Points  \
2         38  OAS30027  OAS30027_MR_d2394                    164   
5         58  OAS30043  OAS30043_MR_d0145                    164   
7         71  OAS30051  OAS30051_MR_d1286                    164   
10        98  OAS30076  OAS30076_MR_d1735                    164   
11       108  OAS30085  OAS30085_MR_d1566                    164   

    Number of ROIs  Gender  MMSE  Age (at imaging)  CDR  CDR-SB  ...  \
2              132    male    26                75  0.5     1.5  ...   
5              132    male    24                86  1.0     5.0  ...   
7              132  female    30                69  0.5     3.0  ...   
10             132  female    23                76  0.5     3.5  ...   
11             132    male    30                70  0.5     3.0  ...   

   Disinhibition  Disinh_sev  Irritability  Irrit_sev Motor Disturbance  \
2            0.0         0.0           0.0        0.0               0.0   
5            0.0         0.0           1.0        2.0               1.0   
7            0.0         0.0           0.0        0.0               0.0   
10           0.0         0.0           0.0        0.0               0.0   
11           1.0         2.0           0.0        0.0               0.0   

    motdist_sev  nighttime behaviors  night_sev  appetite  appeti_sev  
2           0.0                  0.0        0.0       0.0         0.0  
5           1.0                  1.0        1.0       1.0         1.0  
7           0.0                  0.0        0.0       0.0         0.0  
10          0.0                  0.0        0.0       1.0         1.0  
11          0.0                  0.0        0.0       1.0         2.0  

[5 rows x 40 columns]

In [19]:
# Sorted by SUBJECT# 
label_df.sort_values(by=['SUBJECT#'], ascending=True, inplace=True)

In [20]:
label_df = label_df.reset_index(drop=True)

In [21]:
label_df

SUBJECT#   Subject              MR ID  Number of Time Points  \
0            1  OAS30001  OAS30001_MR_d0757                    164   
1            2  OAS30001  OAS30001_MR_d3132                    164   
2            3  OAS30002  OAS30002_MR_d2340                    164   
3            4  OAS30002  OAS30002_MR_d2345                    164   
4            5  OAS30003  OAS30003_MR_d1631                    164   
...        ...       ...                ...                    ...   
1212      1322  OAS31168  OAS31168_MR_d2526                    164   
1213      1323  OAS31169  OAS31169_MR_d0620                    164   
1214      1324  OAS31170  OAS31170_MR_d2410                    164   
1215      1325  OAS31172  OAS31172_MR_d0407                    164   
1216      1326  OAS31172  OAS31172_MR_d1717                    164   

      Number of ROIs  Gender  MMSE  Age (at imaging)  CDR  CDR-SB  ...  \
0                132  female    30                67  0.0     0.0  ...   
1                132  female    30                73  0.0     0.0  ...   
2                132    male    30                73  0.0     0.0  ...   
3                132    male    30                73  0.0     0.0  ...   
4                132  female    30                63  0.0     0.0  ...   
...              ...     ...   ...               ...  ...     ...  ...   
1212             132    male    29                72  0.0     0.0  ...   
1213             132  female    29                64  0.0     0.0  ...   
1214             132  female    25                71  1.0     4.5  ...   
1215             132  female    30                72  0.0     0.0  ...   
1216             132  female    30                75  0.0     0.0  ...   

     Disinhibition  Disinh_sev  Irritability  Irrit_sev Motor Disturbance  \
0              0.0         0.0           0.0        0.0               0.0   
1              0.0         0.0           0.0        0.0               0.0   
2              0.0         0.0           0.0        0.0               0.0   
3              0.0         0.0           0.0        0.0               0.0   
4              0.0         0.0           0.0        0.0               0.0   
...            ...         ...           ...        ...               ...   
1212           0.0         0.0           0.0        0.0               0.0   
1213           0.0         0.0           0.0        0.0               0.0   
1214           0.0         0.0           0.0        0.0               0.0   
1215           0.0         0.0           0.0        0.0               0.0   
1216           0.0         0.0           0.0        0.0               0.0   

      motdist_sev  nighttime behaviors  night_sev  appetite  appeti_sev  
0             0.0                  0.0        0.0       0.0         0.0  
1             0.0                  0.0        0.0       0.0         0.0  
2             0.0                  0.0        0.0       0.0         0.0  
3             0.0                  0.0        0.0       0.0         0.0  
4             0.0                  0.0        0.0       0.0         0.0  
...           ...                  ...        ...       ...         ...  
1212          0.0                  0.0        0.0       0.0         0.0  
1213          0.0                  0.0        0.0       0.0         0.0  
1214          0.0                  1.0        1.0       1.0         3.0  
1215          0.0                  0.0        0.0       0.0         0.0  
1216          0.0                  0.0        0.0       1.0         1.0  

[1217 rows x 40 columns]

In [22]:
# check the labels
print(label_df['Primary Diagnosis'].unique())
print(label_df.shape)
print(label_df['Primary Diagnosis'].head())

['Cognitively normal' 'AD Dementia']
(1217, 40)
0    Cognitively normal
1    Cognitively normal
2    Cognitively normal
3    Cognitively normal
4    Cognitively normal
Name: Primary Diagnosis, dtype: object


In [23]:
import os
# Save the cleaned dataframe
save_path = 'datasets/oasis_labels_bin_dup.csv'
# if not os.path.exists(save_path):
#     label_df.to_csv(save_path, index=False)

In [24]:
female_mask = (label_df['Gender'] == 'female')
print(type(female_mask))
female_df = label_df[female_mask]
print(female_df[['Gender', 'Primary Diagnosis']].value_counts())

male_mask = (label_df['Gender'] == 'male')
print(type(male_mask))
male_df = label_df[male_mask]
print(male_df[['Gender', 'Primary Diagnosis']].value_counts())

<class 'pandas.core.series.Series'>
Gender  Primary Diagnosis 
female  Cognitively normal    626
        AD Dementia            68
Name: count, dtype: int64
<class 'pandas.core.series.Series'>
Gender  Primary Diagnosis 
male    Cognitively normal    425
        AD Dementia            98
Name: count, dtype: int64


In [25]:
label_df.head(2)

SUBJECT#   Subject              MR ID  Number of Time Points  \
0         1  OAS30001  OAS30001_MR_d0757                    164   
1         2  OAS30001  OAS30001_MR_d3132                    164   

   Number of ROIs  Gender  MMSE  Age (at imaging)  CDR  CDR-SB  ...  \
0             132  female    30                67  0.0     0.0  ...   
1             132  female    30                73  0.0     0.0  ...   

  Disinhibition  Disinh_sev  Irritability  Irrit_sev Motor Disturbance  \
0           0.0         0.0           0.0        0.0               0.0   
1           0.0         0.0           0.0        0.0               0.0   

   motdist_sev  nighttime behaviors  night_sev  appetite  appeti_sev  
0          0.0                  0.0        0.0       0.0         0.0  
1          0.0                  0.0        0.0       0.0         0.0  

[2 rows x 40 columns]

In [26]:
# The uniqueness of patients
print(len(label_df['SUBJECT#'].unique()))
print(len(label_df['Subject'].unique()))
print(len(hc_df['Subject'].unique()))

1217
804
660


In [27]:
804 -660

144

### Highlight
1. Currently we have not process the duplicated record of the same subject/patients. If we process it, then there are 660 healthy people and 144 AD patients. \[1051 HC, 166 AD\]     
2. Currently we allow some records to have the empty entries. But they at least have the diagnosis label.    
3. How to process the duplicated records?   
    a. To filter out the record with empty entries on the required attributes like gender?
    b. To pick the data with latest time (larger age?)

#### Processing Duplicate Values

In [28]:
# Check gender value
print(label_df['Gender'].value_counts())
print(label_df['Gender'].isnull().sum()) # Non empty

Gender
female    694
male      523
Name: count, dtype: int64
0


In [29]:
# Sorted by SUBJECT# 
label_df.sort_values(by=['SUBJECT#', 'Age (at imaging)'], ascending=True, inplace=True)
label_df = label_df.reset_index(drop=True)

In [30]:
label_df.head(3)

SUBJECT#   Subject              MR ID  Number of Time Points  \
0         1  OAS30001  OAS30001_MR_d0757                    164   
1         2  OAS30001  OAS30001_MR_d3132                    164   
2         3  OAS30002  OAS30002_MR_d2340                    164   

   Number of ROIs  Gender  MMSE  Age (at imaging)  CDR  CDR-SB  ...  \
0             132  female    30                67  0.0     0.0  ...   
1             132  female    30                73  0.0     0.0  ...   
2             132    male    30                73  0.0     0.0  ...   

  Disinhibition  Disinh_sev  Irritability  Irrit_sev Motor Disturbance  \
0           0.0         0.0           0.0        0.0               0.0   
1           0.0         0.0           0.0        0.0               0.0   
2           0.0         0.0           0.0        0.0               0.0   

   motdist_sev  nighttime behaviors  night_sev  appetite  appeti_sev  
0          0.0                  0.0        0.0       0.0         0.0  
1          0.0                  0.0        0.0       0.0         0.0  
2          0.0                  0.0        0.0       0.0         0.0  

[3 rows x 40 columns]

In [31]:
dup_cols = ['Subject', 'Primary Diagnosis']
label_df = label_df.drop_duplicates(subset=dup_cols, keep='last')
print(label_df.shape)

# Save the cleaned dataframe
save_path = 'datasets/oasis_labels_bin.csv'
# if not os.path.exists(save_path):
#     label_df.to_csv(save_path, index=False)

(815, 40)


In [32]:
print(label_df['Primary Diagnosis'].value_counts())  # Some subject was once normal but became AD later
print(label_df['Gender'].value_counts())

Primary Diagnosis
Cognitively normal    660
AD Dementia           155
Name: count, dtype: int64
Gender
female    459
male      356
Name: count, dtype: int64


In [33]:
print(len(label_df['SUBJECT#'].unique()))

815


## Test for gender group

In [34]:
label_df.head()

SUBJECT#   Subject              MR ID  Number of Time Points  \
1          2  OAS30001  OAS30001_MR_d3132                    164   
3          4  OAS30002  OAS30002_MR_d2345                    164   
6          7  OAS30003  OAS30003_MR_d3731                    164   
10        11  OAS30004  OAS30004_MR_d3457                    164   
14        15  OAS30005  OAS30005_MR_d3453                    164   

    Number of ROIs  Gender  MMSE  Age (at imaging)  CDR  CDR-SB  ...  \
1              132  female    30                73  0.0     0.0  ...   
3              132    male    30                73  0.0     0.0  ...   
6              132  female    30                68  0.0     0.0  ...   
10             132  female    30                64  0.0     0.0  ...   
14             132  female    27                57  0.0     0.0  ...   

   Disinhibition  Disinh_sev  Irritability  Irrit_sev Motor Disturbance  \
1            0.0         0.0           0.0        0.0               0.0   
3            0.0         0.0           0.0        0.0               0.0   
6            0.0         0.0           0.0        0.0               0.0   
10           0.0         0.0           0.0        0.0               0.0   
14           0.0         0.0           0.0        0.0               0.0   

    motdist_sev  nighttime behaviors  night_sev  appetite  appeti_sev  
1           0.0                  0.0        0.0       0.0         0.0  
3           0.0                  0.0        0.0       0.0         0.0  
6           0.0                  0.0        0.0       0.0         0.0  
10          0.0                  0.0        0.0       0.0         0.0  
14          0.0                  0.0        0.0       0.0         0.0  

[5 rows x 40 columns]

In [35]:
y_df = label_df['Primary Diagnosis']
y_df

1       Cognitively normal
3       Cognitively normal
6       Cognitively normal
10      Cognitively normal
14      Cognitively normal
               ...        
1209    Cognitively normal
1212    Cognitively normal
1213    Cognitively normal
1214           AD Dementia
1216    Cognitively normal
Name: Primary Diagnosis, Length: 815, dtype: object

In [36]:
y_df = label_df['Primary Diagnosis'].to_frame()
y_df = y_df.rename({'Primary Diagnosis': 'Label'}, axis='columns')
y_df

Label
1     Cognitively normal
3     Cognitively normal
6     Cognitively normal
10    Cognitively normal
14    Cognitively normal
...                  ...
1209  Cognitively normal
1212  Cognitively normal
1213  Cognitively normal
1214         AD Dementia
1216  Cognitively normal

[815 rows x 1 columns]

In [37]:
gender_df = label_df['Gender'].to_frame() # ['Label', 'Gender']
gender_df

Gender
1     female
3       male
6     female
10    female
14    female
...      ...
1209    male
1212    male
1213  female
1214  female
1216  female

[815 rows x 1 columns]

In [38]:
frames = [y_df, gender_df]
result = pd.concat(frames, axis=1, join='inner')
result.head()

Label  Gender
1   Cognitively normal  female
3   Cognitively normal    male
6   Cognitively normal  female
10  Cognitively normal  female
14  Cognitively normal  female

In [39]:
result

Label  Gender
1     Cognitively normal  female
3     Cognitively normal    male
6     Cognitively normal  female
10    Cognitively normal  female
14    Cognitively normal  female
...                  ...     ...
1209  Cognitively normal    male
1212  Cognitively normal    male
1213  Cognitively normal  female
1214         AD Dementia  female
1216  Cognitively normal  female

[815 rows x 2 columns]

In [49]:
result

Label  Gender
1     Cognitively normal  female
3     Cognitively normal    male
6     Cognitively normal  female
10    Cognitively normal  female
14    Cognitively normal  female
...                  ...     ...
1209  Cognitively normal    male
1212  Cognitively normal    male
1213  Cognitively normal  female
1214         AD Dementia  female
1216  Cognitively normal  female

[815 rows x 2 columns]

In [64]:
result=result.reset_index(drop=True)
result['Label'] = 0
print(type(result.iloc[0]['Label']))
print(type(result.iloc[0]['Gender']))
print(result.iloc[0]['Label'])
print(result.iloc[0]['Gender'])

<class 'numpy.int64'>
<class 'str'>
0
female


In [69]:
test = result['Gender']
print(test.iloc[0])
print(type(test.iloc[0]))

female
<class 'str'>


In [65]:
print(result['Label'])
import torch
labels = result['Label'].to_numpy(dtype='int64')
y = torch.tensor(labels).long().flatten()
y

0      0
1      0
2      0
3      0
4      0
      ..
810    0
811    0
812    0
813    0
814    0
Name: Label, Length: 815, dtype: int64


tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [66]:
print(y[0])
print(type(y[0]))

tensor(0)
<class 'torch.Tensor'>


**Test**

In [40]:
hc_mask = (label_df['Primary Diagnosis'] == 'Cognitively normal')

In [41]:
print(type(hc_mask))
print(label_df[hc_mask].index)
print(type(label_df[hc_mask].index))

<class 'pandas.core.series.Series'>
Index([   1,    3,    6,   10,   14,   15,   17,   19,   22,   23,
       ...
       1202, 1203, 1204, 1205, 1207, 1208, 1209, 1212, 1213, 1216],
      dtype='int64', length=660)
<class 'pandas.core.indexes.base.Index'>


In [42]:
label_df.shape

(815, 40)

In [43]:
ad_mask = (label_df['Primary Diagnosis'] == 'AD Dementia')
print(len(hc_mask))
print(len(ad_mask))

815
815


In [44]:
print(label_df[hc_mask].index)

Index([   1,    3,    6,   10,   14,   15,   17,   19,   22,   23,
       ...
       1202, 1203, 1204, 1205, 1207, 1208, 1209, 1212, 1213, 1216],
      dtype='int64', length=660)


In [45]:
print(label_df[ad_mask].index)

Index([  35,   53,   65,   90,  100,  101,  105,  109,  113,  125,
       ...
       1088, 1089, 1096, 1121, 1134, 1135, 1149, 1171, 1179, 1214],
      dtype='int64', length=155)


In [46]:
label_df[['Subject', 'Gender', 'Primary Diagnosis']].head

<bound method NDFrame.head of        Subject  Gender   Primary Diagnosis
1     OAS30001  female  Cognitively normal
3     OAS30002    male  Cognitively normal
6     OAS30003  female  Cognitively normal
10    OAS30004  female  Cognitively normal
14    OAS30005  female  Cognitively normal
...        ...     ...                 ...
1209  OAS31167    male  Cognitively normal
1212  OAS31168    male  Cognitively normal
1213  OAS31169  female  Cognitively normal
1214  OAS31170  female         AD Dementia
1216  OAS31172  female  Cognitively normal

[815 rows x 3 columns]>

In [47]:
female_mask = (label_df['Gender'] == 'female')
print(type(female_mask))
female_df = label_df[female_mask]
print(female_df[['Gender', 'Primary Diagnosis']].value_counts())

<class 'pandas.core.series.Series'>
Gender  Primary Diagnosis 
female  Cognitively normal    393
        AD Dementia            66
Name: count, dtype: int64


In [48]:
male_mask = (label_df['Gender'] == 'male')
print(type(male_mask))
male_df = label_df[male_mask]
print(male_df[['Gender', 'Primary Diagnosis']].value_counts())

<class 'pandas.core.series.Series'>
Gender  Primary Diagnosis 
male    Cognitively normal    267
        AD Dementia            89
Name: count, dtype: int64
